combinetf2/fitter.py

```diff
@@ -0,0 +1,1410 @@
+        )
+
+    def toyassign(self, bayesian=False, bootstrap_data=False):
+        if bayesian:
+            # randomize actual values
+            self.bayesassign()
+        else:
+            # randomize nuisance constraint minima
+            self.frequentistassign()
+
+        if self.binByBinStat:
+            # TODO properly implement randomization of constraint parameters associated with bin-by-bin stat nuisances for frequentist toys,
+            # currently bin-by-bin stat fluctuations are always handled in a bayesian way in toys
+            # this also means bin-by-bin stat fluctuations are not consistently propagated for bootstrap toys from data
+            self.beta0.assign(
+                tf.random.gamma(
+                    shape=[],
+                    alpha=self.indata.kstat + 1.0,
+                    beta=self.indata.kstat,
+                    dtype=self.beta0.dtype,
+                )
+            )
+
+        if bootstrap_data:
+            # randomize from observed data
+            if self.binByBinStat and not bayesian:
+                raise Exception(
+                    "Since bin-by-bin statistical uncertainties are always propagated in a bayesian manner, \
+                    they cannot currently be consistently propagated for bootstrap toys"
+                )
+            self.nobs.assign(
+                tf.random.poisson(
+                    lam=self.indata.data_obs, shape=[], dtype=self.nobs.dtype
+                )
+            )
+        else:
+            # randomize from expectation
+            self.nobs.assign(
+                tf.random.poisson(
+                    lam=self.expected_yield(),
+                    shape=[],
+                    dtype=self.nobs.dtype,
+                )
+            )
+
+        # assign start values for nuisance parameters to constraint minima
+        self.xdefaultassign()
+        # set likelihood offset
+        self.nexpnom.assign(self.expected_yield())
+
+    def _compute_impact_group(self, v, idxs):
+        cov_reduced = tf.gather(self.cov[self.npoi :, self.npoi :], idxs, axis=0)
+        cov_reduced = tf.gather(cov_reduced, idxs, axis=1)
+        v_reduced = tf.gather(v, idxs, axis=1)
+        invC_v = tf.linalg.solve(cov_reduced, tf.transpose(v_reduced))
+        v_invC_v = tf.einsum("ij,ji->i", v_reduced, invC_v)
+        return tf.sqrt(v_invC_v)
+
+    @tf.function
+    def impacts_parms(self):
+        # impact for poi at index i in covariance matrix from nuisance with index j is C_ij/sqrt(C_jj) = <deltax deltatheta>/sqrt(<deltatheta^2>)
+        cov_poi = self.cov[: self.npoi]
+        cov_noi = tf.gather(self.cov[self.npoi :], self.indata.noigroupidxs)
+        v = tf.concat([cov_poi, cov_noi], axis=0)
+        impacts = v / tf.reshape(tf.sqrt(tf.linalg.diag_part(self.cov)), [1, -1])
+
+        # impact data stat # FIXME: This might not be correct in case noi != systsnoconstraint
```

Comment:
I don't think the FIXME comment is relevant, noi's are irrelevant for this purpose